### ISSUE TITLE
Docker leaks container netns netlink sockets to forked docker-proxy processes, preventing netns cleanup after container is stopped

### ISSUE THREAD

**Issue by SpComb** [docker/docker#32090]
I have been debugging a networking issue when using Weave networking, where Weave overlay networking addresses associated with Docker containers continue to respond to pings even after the Docker container has been destroyed. This causes issues when the Weave overlay networking address is re-allocated for use by a different container, as both the legit container and the undead zombie container continue to respond got ARP queries for the IP address. Clients are randomly unable to connect to services when their traffic ends up at the zombie container: https://github.com/weaveworks/weave/issues/2842 https://github.com/weaveworks/weave/issues/2433#issuecomment-265993860

This is fundamentally caused by leaking references to the container's netns. When a Docker container stops, Docker will explicitly delete its own `eth0` veth interface pair before unmounting the `/run/docker/netns/*`. For the Weave `ethwe` veth pair, we rely on the interfaces being destroyed by the kernel when the last reference to the container's network namespace goes away. Unfortunately, there are multiple cases where these network namespace references can leak, meaning that the container's network namespace does not get destroyed by the kernel after Docker considers the container to be destroyed. With normal Docker networking, these empty network namespaces are harmless, but when using Weave networking, the containers turn into ghosts that continue to haunt the network :ghost: 

So far, I have found two cases of this happening:

* https://github.com/kontena/kontena/issues/2004 Running cadvisor in a Docker container with the default `rprivate` mode for the `-v /:/rootfs` bind-mount will capture any `/run/docker/netns/*` mounts that exist when the cadvisor container is started, and these may ever not go away when Docker later unlinks the `/run/docker/netns/*` mountpoint, depending on the kernel version
* This issue, where `dockerd` launching a new `docker-proxy` process captures references to any running Docker container network namspaces in the form of netlink sockets inherited from the forking parent

Please refer to my earlier issue comments for further background on this issue:

* https://github.com/weaveworks/weave/issues/2433#issuecomment-265993860
* https://github.com/kontena/kontena/issues/2004
* https://gist.github.com/SpComb/9cc2a7ed46ff708547d09854b78c197f#weave-zombie-huntersh

And the debugging for this specific issue:
* https://github.com/weaveworks/weave/issues/2842#issuecomment-288054662 onwards
* https://gist.github.com/SpComb/f83c53b05aad79c6213185a3da7a7902
* https://gist.github.com/SpComb/197cf7c4191dcca261fdfeade0be3c54

## Description

It seems that the `dockerd` process holds open a netlink socket for each running Docker container, associated with the container's network namespace. It also seems that when Docker forks a new `docker-proxy` process, these netlink socket FDs are inherited by the `docker-proxy` process: https://github.com/weaveworks/weave/issues/2842#issuecomment-289022778

After Docker stops a container and closes its own netlink socket, that container's network namespace will not be cleaned up by the kernel until the `docker-proxy` processes for some unrelated container exit, and these leaked netlink sockets are closed by the kernel: https://github.com/weaveworks/weave/issues/2842#issuecomment-288791826

## Steps to reproduce the issue

See the gist: https://gist.github.com/SpComb/c8604792fc9d3ba0518d889ff8f89ccb

1. `git clone https://gist.github.com/SpComb/c8604792fc9d3ba0518d889ff8f89ccb docker-netns-leak && cd docker-netns-leak`
2. `sudo ./test.sh`

Run a `victim` container, and attach an extra veth pair connected to a `br-test` bridge, configured with a `192.168.0.10` address.

Run a second `bystander` container, using `-p 6379:6379` to publish a port. A new `docker-proxy` process is forked, which inherits the netlink socket from the `victim` network namespace.

The `victim` container is stopped, and the extra veth pair does not get destroyed, and the `192.168.0.10` address inside the stopped container continues to respond to pings.

This does not happen unless the `docker-proxy` process is started while the `victim` container is running, and the extra veth pair from the `victim` container goes away once the `docker-proxy` process is stopped.

## Describe the results you received

Stopping the `victim` container only destroys the Docker veth pair:

```
80: veth102a2ef@veth84ba9ce: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
81: veth84ba9ce@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue master docker0 state DOWN group default 
    link/ether 6e:25:9f:c2:3c:9e brd ff:ff:ff:ff:ff:ff
Deleted 80: veth102a2ef@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 81: veth84ba9ce@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether 6e:25:9f:c2:3c:9e
```

The extra `veth` pair attached to the stopped `victim` container's network namespace does not get destroyed, while the `bystander` container is still running with its `docker-proxy` process:

```
83: veth0@if82: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-test state UP mode DEFAULT group default qlen 1000
    link/ether b2:3c:93:87:53:20 brd ff:ff:ff:ff:ff:ff
85: veth8897a0a@if84: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP mode DEFAULT group default 
    link/ether c6:f9:27:97:fb:b3 brd ff:ff:ff:ff:ff:ff

```

It also continues to responds to pings:

```
PING 192.168.0.10 (192.168.0.10) 56(84) bytes of data.
64 bytes from 192.168.0.10: icmp_seq=1 ttl=64 time=0.118 ms
64 bytes from 192.168.0.10: icmp_seq=2 ttl=64 time=0.148 ms
```

Stopping the `bystander` container will then stop the `docker-proxy` process, and the `victim` container's extra veth pair gets destroyed, as well as the `bystander` container's Docker veth pair:

```
Deleted 83: veth0@NONE: <BROADCAST,MULTICAST> mtu 1500 master br-test state DOWN 
    link/ether b2:3c:93:87:53:20

84: vethbf7273c@veth8897a0a: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
Deleted 84: vethbf7273c@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
Deleted 85: veth8897a0a@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether c6:f9:27:97:fb:b3
```

## Describe the results you expected

If the `bystander` container with its associated `docker-proxy` is not running, then stopping the `victim` container will correctly destroy the Docker veths, release the network namespace and thus also cause the extra veths to be destroyed immediately:

```
76: veth91821c3@veth0b65d5c: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 76: veth91821c3@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 77: veth0b65d5c@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether 4e:c3:f5:87:c8:dd

Deleted 79: veth0@NONE: <BROADCAST,MULTICAST> mtu 1500 master br-test state DOWN 
    link/ether 8e:44:24:a5:c3:67
```

## Output of `docker version`:

```
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue Mar 14 21:10:49 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue Mar 14 21:10:49 2017
 OS/Arch:      linux/amd64
```

## Output of `docker info`:

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.12.6
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp selinux
Kernel Version: 4.9.9-coreos-r1
Operating System: Container Linux by CoreOS 1298.6.0 (Ladybug)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.9 MiB
Name: coreos-1gb-ams2-01
ID: HLW2:MFQE:MQHF:UWLC:R6VH:E2HW:7GG2:WEMD:DZSI:GSGS:A5HA:HK5E
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

## Additional environment details

CoreOS stable 1298.6.0 on DigitalOcean.

**Comment by cpuguy83** [comment]
Wow, thanks for the detailed analysis!

Ping @docker/core-libnetwork-maintainers

**Comment by wenjianhn** [comment]
> It seems that the dockerd process holds open a netlink socket for each running Docker container, associated with the container's network namespace. It also seems that when Docker forks a new docker-proxy process, these netlink socket FDs are inherited by the docker-proxy process: weaveworks/weave#2842 (comment)

I think your guess is right.The node ids are the same.
They are 16185 and 16186.
> [root@localhost ~]# lsof -p 18680 |grep netlink
docker-pr 18680 root    9u  netlink             0t0      16185 ROUTE
docker-pr 18680 root   10u  netlink             0t0      16186 XFRM
[root@localhost ~]# lsof -p 710 |grep netlink
dockerd 710 root    9u  netlink                         0t0      16185 ROUTE
dockerd 710 root   10u  netlink                         0t0      16186 XFRM

**Comment by cpuguy83** [comment]
Originally was thinking this may be due to https://github.com/vishvananda/netns/blob/master/netns_linux.go#L71 not using `O_CLOEXEC`, however I still see docker-proxy getting leaked sockets even with setting it.

You can see this with:
```go
# docker run -d --name test` -p 80 busybox top
# docker run -d --name test2 -p 80 busybox top
```

The `docker-proxy` for the process for the `test1` container will have exactly 3 socket fd's open,  for `test2` there will be 4 socket fd's, if you do another 5, so on...

If you do `lsof | grep sock`, the first docker-proxy will have  entries, 2nd docker-proxy will have a handle to 1, if you do a 3rd it will have 2 handles, etc.

Checking with `/proc/net/netlink` the 2nd docker-proxy definitely has the netlink handle, and each new docker-proxy inherits all existing netlink handles.

**Comment by mavenugo** [comment]
@SpComb Thanks for the detailed report.

@aboch ^^

**Comment by wenjianhn** [comment]
@aboch The patch works.
No more netlink sockets for the docker-proxy.

> [root@localhost docker]# lsof -p 15485
COMMAND     PID USER   FD      TYPE DEVICE SIZE/OFF       NODE NAME
docker-pr 15485 root  cwd       DIR  253,0     4096      15752 /go/src/github.com/docker/docker
docker-pr 15485 root  rtd       DIR  0,103      230        256 /
docker-pr 15485 root  txt       REG  0,103  2530685      83319 /usr/local/bin/docker-proxy
docker-pr 15485 root  mem       REG   0,35               83319 /usr/local/bin/docker-proxy (stat: No such file or directory)
docker-pr 15485 root    0r      CHR    1,3      0t0     122397 /dev/null
docker-pr 15485 root    1w      CHR    1,3      0t0     122397 /dev/null
docker-pr 15485 root    2w      CHR    1,3      0t0     122397 /dev/null
docker-pr 15485 root    4u     sock    0,7      0t0     130509 protocol: TCPv6
docker-pr 15485 root    5u  a_inode    0,9        0       5832 [eventpoll]
docker-pr 15485 root   10r      REG  0,107        0 4026532322 net

> [root@localhost docker]# lsof -p 15560
COMMAND     PID USER   FD      TYPE DEVICE SIZE/OFF       NODE NAME
docker-pr 15560 root  cwd       DIR  253,0     4096      15752 /go/src/github.com/docker/docker
docker-pr 15560 root  rtd       DIR  0,103      230        256 /
docker-pr 15560 root  txt       REG  0,103  2530685      83319 /usr/local/bin/docker-proxy
docker-pr 15560 root  mem       REG   0,35               83319 /usr/local/bin/docker-proxy (stat: No such file or directory)
docker-pr 15560 root    0r      CHR    1,3      0t0     122397 /dev/null
docker-pr 15560 root    1w      CHR    1,3      0t0     122397 /dev/null
docker-pr 15560 root    2w      CHR    1,3      0t0     122397 /dev/null
docker-pr 15560 root    4u     sock    0,7      0t0     128226 protocol: TCPv6
docker-pr 15560 root    5u  a_inode    0,9        0       5832 [eventpoll]
docker-pr 15560 root   10r      REG  0,107        0 4026532322 net

**Comment by SpComb** [comment]
Built from the 1.12.x branch, and confirmed the [`test.sh`](https://gist.github.com/SpComb/c8604792fc9d3ba0518d889ff8f89ccb#file-test-sh) repro case.

Patched in https://github.com/vishvananda/netlink/pull/209 and confirmed that this fixes the repro case:

```
+ docker stop victim
victim
+ echo > The victim has been stopped, and the veth should be gone...
> The victim has been stopped, and the veth should be gone...
+ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                              PORTS                    NAMES
60feaf059e8a        redis               "docker-entrypoint.sh"   4 seconds ago       Up 3 seconds                        0.0.0.0:6379->6379/tcp   bystander
c17089236328        redis               "docker-entrypoint.sh"   7 seconds ago       Exited (0) Less than a second ago                            victim
+ ip link show veth0
Device "veth0" does not exist.
+ ping -c 2 192.168.0.10
PING 192.168.0.10 (192.168.0.10): 56 data bytes
92 bytes from 192.168.0.1: Destination Host Unreachable
92 bytes from 192.168.0.1: Destination Host Unreachable
--- 192.168.0.10 ping statistics ---
2 packets transmitted, 0 packets received, 100% packet loss
+ grep NETLINK
+ pidof docker-proxy
+ lsof -p 12791
```

### `strace` of `dockerd`
Confirmed to be using `SOCK_CLOEXEC`:
```
[pid 12251] socket(PF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 23
```

**Comment by SpComb** [comment]
Also tested as fixed in `docker-ce=17.05.0~ce-0~ubuntu-xenial`
